cpp/src/arrow/adapters/orc/util.cc

```diff
@@ -1111,7 +1120,11 @@ Result<std::shared_ptr<DataType>> GetArrowType(const liborc::Type* type) {
     case liborc::CHAR:
       return fixed_size_binary(static_cast<int>(type->getMaximumLength()));
     case liborc::TIMESTAMP:
+      // The timestamp values stored in ORC are in the writer timezone.
       return timestamp(TimeUnit::NANO);
```

Comment:
Is this true? I would have expected the values in the "UTC" timezone and the expectations is readers should interpret it in the local time zone.

Reply:
Yes, you're right.

For `orc::TIMESTAMP` type:
- The Orc writer expects input data (i.e. in the orc::TimestampVectorBatch) to be in the "UTC" timezone and serializes it into the writer timezone: https://github.com/apache/orc/blob/main/c%2B%2B/src/ColumnWriter.cc#L1717
- The Orc reader deserializes the data from writer timezone and restores it into reader timezone: https://github.com/apache/orc/blob/main/c%2B%2B/src/ColumnReader.cc#L336

For `orc::TIMESTAMP_INSTANT` type:
- The Orc writer expects input data to be in the "UTC" timezone and serializes it into the "UTC" timezone: https://github.com/apache/orc/blob/main/c%2B%2B/src/ColumnWriter.cc#L1644
- The Orc reader deserializes the data from "UTC" timezone and no more conversion is needed because writerTimezone and readerTimezone are both "UTC": https://github.com/apache/orc/blob/main/c%2B%2B/src/ColumnReader.cc#L282

We have seen many issues around `orc::TIMESTAMP` type because of the writer-reader timezone conversion, especially with different day-light saving rules. So that's why `orc::TIMESTAMP_INSTANT` type is added and is always preferred over `orc::TIMESTAMP` type if user can take care of the timezone.

Reply:
Okay that makes sense. It does make me wonder if we should then be applying the local time zone to these types, instead of using naive timestamps:

```suggestion
      // The timestamp values stored in ORC are in the writer timezone.
      return timestamp(TimeUnit::NANO, GetLocalTimeZone());
```

Reply:
This would fail the equality check of timestamp types in the roundtrip tests. Additionally, it seems that there is a lack of utility like `GetLocalTimeZone()`?

IMHO, if writer uses naive timestamp types, reader should follow the same pattern.